### PR TITLE
Feature/any

### PR DESCRIPTION
**Pull request by bcumming**
Partial implementation of `std::any` from C++17 standard.
     http://en.cppreference.com/w/cpp/utility/any

The implementation is in the `util` library as `util::any`.

Implements a standard-compliant subset of the full interface.

  * Does not avoid dynamic allocation of small objects.
  * Does not implement the `in_place_type<T>` constructors from the standard.
  * Does not implement the `emplace` modifier from the standard.

### CODE REVIEW

File: src/util/any.hpp [hunk 1]
```diff
@@ -0,0 +1,269 @@
...
+
+    template <typename ValueType, typename U=any_cast_remove_qual<ValueType>>
+    using test_any_cast_constructable = typename
+        std::conditional<
+               std::is_constructible<ValueType, const U&>::value
+            && std::is_constructible<ValueType, U&>::value
+            && std::is_constructible<ValueType, U>::value,
+            std::true_type,
+            std::false_type>::type;
+
```
Comment (halfflat) [inline src/util/any.hpp]: That said, this test seems too specific. Looking at the reference, the requirements on `ValueType` depend on the particular overload of `any_cast`.

File: src/util/any.hpp [hunk 2]
```diff
@@ -0,0 +1,269 @@
...
+        model(const T& other):
+            value(other)
+        {}
+
+        model(T&& other):
+            value(std::move(other))
+        {}
+
+        void* pointer() {
```
Comment (halfflat) [inline src/util/any.hpp]: missing `override`

File: src/util/any.hpp [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+            return &value;
+        }
+
+        const void* pointer() const {
```
Comment (halfflat) [inline src/util/any.hpp]: missing `override`

File: src/util/any.hpp [hunk 4]
```diff
@@ -0,0 +1,269 @@
...
+        impl::test_any_cast_constructable<T>::value,
+        "any_cast requires ");
+
+    if (operand==nullptr || !operand->has_value()) {
+        return nullptr;
+    }
+
+    if (operand->type() != typeid(T)) {
+        throw bad_any_cast();
+    }
```
Comment (halfflat) [inline src/util/any.hpp]: The `any_cast<T>(const any*)` overload shouldn't throw (in fact, should be no-except). When `typeid(T)` is not the same as `operand->type()`, it should return `nullptr`.

As an implementation note, you could potentially replace the private `unsafe_cast<T>` method with a 'safe' version that did the `typeid(T)` test and returned `nullptr` on failure.

This applies of course to the non-const version below.

File: src/util/any.hpp [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        std::conditional<
+            std::is_move_constructible<T>::value
+            & !std::is_lvalue_reference<T>::value,
+            std::true_type, std::false_type>::type;
+
+    auto ptr = any_cast<U>(&operand);
+    if (ptr==nullptr) {
+        throw bad_any_cast();
+    }
+    return impl::move_else_copy<T>(static_cast<T*>(ptr), do_move());
```
Comment (halfflat) [inline src/util/any.hpp]: This looks fairly complicated (in terms of conditional moving etc.). Does `return std::move(*any_cast<T>(&operand));` not work? If not, an explicatory comment would help.

File: src/util/any.hpp [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    void swap(any& other) noexcept {
+        std::swap(other.state_, state_);
+    }
+
+    bool has_value() const noexcept {
+        return (bool)state_;
+    }
+
+    const std::type_info& type() const noexcept {
+        return state_->type();
```
Comment (halfflat) [inline src/util/any.hpp]: Should return `typeid(void)` if `!has_value()`. (This may simplify the `any_cast` implementations below, too.)

File: tests/unit/test_any.cpp [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+    util::any ai(42);
+    auto i = util::any_cast<int>(ai);
+    EXPECT_EQ(typeid(i), typeid(int));
+    EXPECT_EQ(i, 42);
+}
+
+// test any_cast(any&&)
+TEST(any, any_cast_rvalue) {
+    auto moved = util::any_cast<moveable>(util::any(moveable()));
+    std::cout << "moves and copies: " << moved.moves << " " << moved.copies  << "\n";
```
Comment (halfflat) [inline tests/unit/test_any.cpp]: This test looks like it might be still being prototyped.

File: src/util/any.hpp [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+
+    template <typename T>
+    friend const T* any_cast(const any* operand);
+
+    template <typename T>
+    friend T* any_cast(any* operand);
+
+    template <typename T>
+    T* unsafe_cast() {
+        return reinterpret_cast<T*>(state_->pointer());
```
Comment (bcumming) [inline src/util/any.hpp]: You are right. This was a hang over from a slightly different earlier implementation which required a more "industrial" solution.

File: src/util/any.hpp [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+} // namespace impl
+
+// If operand is not a null pointer, and the typeid of the requested T matches
+// that of the contents of operand, a pointer to the value contained by operand,
+// otherwise a null pointer.
+template<class T>
+const T* any_cast(const any* operand) {
+    static_assert(
+        impl::test_any_cast_constructable<T>::value,
+        "any_cast requires ");
```
Comment (bcumming) [inline src/util/any.hpp]: The standard does not say that it is a strict compile time requirement, but it does state that the program is ill-formed if this condition is violated, so I figured that I could just catch the problem at compile time.

File: src/util/any.hpp (hunk 4 above)
Comment (bcumming) [inline src/util/any.hpp]: True that.

File: src/util/any.hpp [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+
+template<class T>
+T any_cast(any&& operand) {
+    using U = impl::any_cast_remove_qual<T>;
+    using do_move = typename
+        std::conditional<
+            std::is_move_constructible<T>::value
+            & !std::is_lvalue_reference<T>::value,
+            std::true_type, std::false_type>::type;
+
```
Comment (bcumming) [inline src/util/any.hpp]: You learn something new every day. Will do.

File: src/util/any.hpp (hunk 6 above)
Comment (bcumming) [inline src/util/any.hpp]: good catch.

Comment (bcumming) [inline src/util/any.hpp]: Fixed, and a unit test for `any::type` was added.

File: src/util/any.hpp (hunk 1 above)
Comment (bcumming) [inline src/util/any.hpp]: Right you are. I got a bit confused when reading the specs, and thought that it was an enumerated list of requirements that applied to all cases. I will narrow the scope of the tests.

File: src/util/any.hpp (hunk 5 above)
Comment (bcumming) [inline src/util/any.hpp]: It does work... I was making life more complicated than it had to be.

File: src/util/any.hpp (hunk 4 above)
Comment (bcumming) [inline src/util/any.hpp]: I have kept the `unsafe_cast`. Moving the `typeid(T)` makes another `if` statement, because we still have to test if `operand` is not null.

File: src/util/any.hpp [hunk 11]
```diff
@@ -0,0 +1,269 @@
...
+    }
+};
+
+class any {
+public:
+    constexpr any() = default;
+
+    any(const any& other):
+        state_(other.state_->copy())
+    {}
```
Comment (bcumming) [inline src/util/any.hpp]: I have tried to make things more compact.

**Comment by bcumming** [comment]
I have not explicitly implemented `std::swap(util::any)`, because we get it for free with the appropriate move assignment.

However the standard explicitly states that it should be overloaded to call the `any::swap` member modifier as follows `lhs.swap(rhs)`

http://en.cppreference.com/w/cpp/utility/any/swap2

So, do I actually have to explicitly overload it?

**Comment by bcumming** [comment]
I also prefer not to have friend functions defined inline inside a class. No strong reason, I just find it easier to read that way.

File: src/util/any.hpp [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+    }
+    return static_cast<T>(std::move(*ptr));
+}
+
+// Constructs an any object containing an object of type T, passing the
+// provided arguments to T's constructor.
+template <class T, class... Args>
+any make_any(Args&&... args) {
+    return any(T(std::forward<Args>(args) ...));
+}
```
Comment (halfflat) [inline src/util/any.hpp]: I believe this is not quite the same as what the standard promises, in that this will construct the `T` object and then (generally speaking) use `T`'s move constructor to place it in the wrapped store, while the standard description indicates that the wrapped `T` be constructed in-place.

Solution would be to add comment documenting the difference (or verifying that there is in fact no difference, but I don't think this is true) or implementing the `in_place_t` constructors and invoking them with `make_any` via perfect forwarding.

Comment (bcumming) [inline src/util/any.hpp]: I added a comment. Let's wait until #256.
